woocommerce/woocommerce-ios
# Mitigation: Custom Amounts CTA not being tappable unless keyboard is manually dismissed

Closes: #13955

## Description
We’ve seen some instances where the Custom Amounts CTA button is not tappable unless the keyboard is manually dismissed first, this issue hasn’t been replicable consistently, but we can infer that either the keyboard layout is interfering with the button tap gesture, or the timing of SwiftUI managing the keyboard vs the view dismissal.

We attempt to mitigate this problem by: 
- Allowing keyboard dismissal when a merchant taps anywhere else within the view. Adding this within the button closure wouldn’t work, as if the button happens to be non-tappable due to the keyboard layout constraints, we wouldn’t be able to call it’s dismissal either.
- Alleviate potential layout conflicts involving the keyboard within SwiftUI by ignoring the keyboard safe area.

## Testing information
As shared in the PR issue, this is cannot be consistently reproduced, but we can run the flow a few times and test the changes:

- On a physical device, start order creation
- Add a few products
- Tap "Add Custom Amount"  > Select any option
- Enter details
- Tap the CTA "Add Custom Amount" to save the changes
- Notice that the tap works normally, the keyboard is dismissed, as well as the view
- Observe that when the keyboard is open, you can tap outside to dismiss it, then tap again within the name text field or the number textfield (limited to where the actual placeholder is) in to make the keyboard appear again.

No new Unit Tests have been added, as it's merely an UI update. 

Since the keyboard dismissal logic is encapsulated to this view only, we've just added as an extension of the view. We can move this to the view model if you prefer, but I opted to keep it in the view since we don't need to expose it to other views through the view model if the problem hasn't been reported anywhere else.

---

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:
